### PR TITLE
jit: support unreachable instruciton, and stack trace.

### PR DESCRIPTION
**Pull request by mathetake**
Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

part of #65

### CODE REVIEW

File: wasm/jit/engine.go [hunk 1]
```diff
@@ -34,20 +35,43 @@ type engine struct {
...
+	defer func() {
+		if v := recover(); v != nil {
+			top := e.callFrameStack
+			var traces []string
+			var counter int
+			for top != prevFrame {
+				traces = append(traces, fmt.Sprintf("\t%d: %s", counter, top.getFunctionName()))
+				top = top.caller
+				counter++
+				// TODO: include DWARF symbols.
```
Comment (mathetake) [inline wasm/jit/engine.go:53]: Add reference to #58 

```suggestion
				// TODO: include DWARF symbols. See #58 
```

File: wasm/jit/engine.go [hunk 2]
```diff
@@ -226,18 +254,34 @@ type callFrame struct {
...
+}
+
+type compiledHostFunction = struct {
+	f    func(ctx *wasm.HostFunctionCallContext)
+	name string
+}
+
 type compiledWasmFunction struct {
+	// FunctionInstance from which this is compiled.
+	originalFunctionInstance *wasm.FunctionInstance
```
Comment (codefromthecrypt) [inline wasm/jit/engine.go:284]: maybe `source` ?

File: wasm/jit/testdata/unreachable.wat [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+		(call $one)
+	)
+	(func $one
+		(call $two)
+	)
+	(func $two
+		(call $three)
+	)
+	(func $three
+		(unreachable)
```
Comment (codefromthecrypt) [inline wasm/jit/testdata/unreachable.wat:12]: is there a more intuitive use of unreachable? like something where it is may occur in real life (not a forged example)

File: wasm/jit/engine.go [hunk 4]
```diff
@@ -34,20 +35,43 @@ type engine struct {
...
 }
 
-type hostFunction = func(ctx *wasm.HostFunctionCallContext)
-
 func (e *engine) Call(f *wasm.FunctionInstance, args ...uint64) (returns []uint64, err error) {
+	prevFrame := e.callFrameStack
+	defer func() {
+		if v := recover(); v != nil {
+			top := e.callFrameStack
+			var traces []string
```
Comment (codefromthecrypt) [inline wasm/jit/engine.go:47]:
```suggestion
			var frames []string
```

File: wasm/jit/engine.go [hunk 5]
```diff
@@ -34,20 +35,43 @@ type engine struct {
...
-	hostFunctionIndex map[*wasm.FunctionInstance]int64
+	compiledHostFunctions     []*compiledHostFunction
+	compiledHostFunctionIndex map[*wasm.FunctionInstance]int64
 }
 
-type hostFunction = func(ctx *wasm.HostFunctionCallContext)
-
 func (e *engine) Call(f *wasm.FunctionInstance, args ...uint64) (returns []uint64, err error) {
+	prevFrame := e.callFrameStack
+	defer func() {
```
Comment (codefromthecrypt) [inline wasm/jit/engine.go:44]: comment why specifically here we are handling panic?

**Review by codefromthecrypt — APPROVED** [review]
as usual, mostly naming and semantics!

File: wasm/jit/testdata/unreachable.wat (hunk 3 above)
Comment (mathetake) [reply wasm/jit/testdata/unreachable.wat:12]: for example `unreachable` will be placed in a branch for the case when the slice boundary check failed (like TinyGO does emit unreachable on "slice out of bounds")!

Comment (mathetake) [reply wasm/jit/testdata/unreachable.wat:12]: But it is essentially the same as this example

**Comment by codefromthecrypt** [comment]
thanks for making it possible to uhh become unreachable!

**Pull request merged**